### PR TITLE
Tool extension host and doctor prototype

### PR DESCRIPTION
**Pull request by jonahwilliams**
### Description

This PR contains a basic implementation of a shared isolate tool extension, as well as the basis for the public api for doctor and an implementation for web and linux doctor.

It also includes shimming logic in flutter doctor to allow us to make a gradual transition. Because the API will differ slightly between the current doctor and the new doctor, not all the types have been migrated yet.

### Serialization

To support a future in which extensions run in different isolates, all returned values must be serializable to JSON. This serialization is implemented here by hand to reduce the scope of dependencies needed and to make sure that the cost of increasing the public api surface is felt immediately by the author.

Because these are not currently running cross isolate, there is an assert which attempts to JSON serialize the result and fails if that can't be done.

### DoctorDomain

This is intended to contain all functionality related to diagnosing problems with the users environment, and potentially project specific problems in the future. It contains a single method, diagnose, which returns a set of validation results.

```
/// Functionality related to diagnosing problems with a project or development
/// environment.
abstract class DoctorDomain extends Domain {

   /// Diagnose problems with the user's environment.
  Future<ValidationResult> diagnose(Map<String, Object> arguments);
}
```

Since we've not reached a period of stabilization with the API, we're expecting to make adjustments here as we move more of the tool to use this API.

In the future, we expect developers to be able to write standalone validators which may be "installed" or otherwise integrated into their flutter tool. This would allow a package which requires sufficient configuration to provide a project specific validator (such as firebase).

### Dependency injection

Since these tool extensions are intended to eventually be public APIs, they don't use the flutter tool DI system. For now, I've provided fields on the domain which points to a `FileSystem`, `ProcessManager`, and `Platform` implementation from the same packages that flutter uses. 

### Import restrictions

The files in `lib/src/extensions` are not allowed to import anything outside of this directory, this is to facilitate moving it to a separate package that is publicly usable

### CODE REVIEW

File: packages/flutter_tools/lib/src/doctor.dart [hunk 1]
```diff
@@ -56,6 +61,15 @@ class _DefaultDoctorValidatorsProvider implements DoctorValidatorsProvider {
   List<DoctorValidator> _validators;
   List<Workflow> _workflows;
 
+  final extension.ExtensionShim extensionShim = extension.ExtensionShim(<extension.ToolExtension>[
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/doctor.dart]: This is a bit of a temporary measure until we can bootstrap something that works in g3 too.

File: packages/flutter_tools/lib/src/extension/doctor.dart [hunk 2]
```diff
@@ -0,0 +1,222 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert';
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/extension/doctor.dart]: Everything in this file will become a public API for the tool

File: packages/flutter_tools/lib/src/extension/doctor.dart [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+  @override
+  String toString() => json.encode(toJson());
+}
+
+/// Functionality related to diagnosing problems with a project or development
+/// environment.
+abstract class DoctorDomain extends Domain {
+
+  /// Diagnose problems with the user's environment.
+  Future<ValidationResult> diagnose(Map<String, Object> arguments);
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/extension/doctor.dart]: Not clear whether we should pass through arbitrary arguments or map it to something typed first. dynamic is better for forwards/backwards compatibility.

Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/extension/doctor.dart]: This should change to a list so that we can return separate validation results.

File: examples/flutter_gallery/ios/Runner.xcodeproj/project.pbxproj [hunk 4]
```diff
@@ -42,6 +42,7 @@
 		2D9222441EC1E1BA007564B0 /* GeneratedPluginRegistrant.m */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = sourcecode.c.objc; path = GeneratedPluginRegistrant.m; sourceTree = "<group>"; };
 		3B3967151E833CAA004F5970 /* AppFrameworkInfo.plist */ = {isa = PBXFileReference; fileEncoding = 4; lastKnownFileType = text.plist.xml; name = AppFrameworkInfo.plist; path = Flutter/AppFrameworkInfo.plist; sourceTree = "<group>"; };
 		3B80C3931E831B6300D905FE /* App.framework */ = {isa = PBXFileReference; lastKnownFileType = wrapper.framework; name = App.framework; path = Flutter/App.framework; sourceTree = "<group>"; };
+		4E9FBE7307263274EFEF7AA8 /* Pods-Runner.release.xcconfig */ = {isa = PBXFileReference; includeInIndex = 1; lastKnownFileType = text.xcconfig; name = "Pods-Runner.release.xcconfig"; path = "Pods/Target Support Files/Pods-Runner/Pods-Runner.release.xcconfig"; sourceTree = "<group>"; };
```
Comment (zanderso) [inline examples/flutter_gallery/ios/Runner.xcodeproj/project.pbxproj]: Are the changes to this file intentional?

File: packages/flutter_tool_api/lib/doctor.dart [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+  @override
+  String toString() => json.encode(toJson());
+}
+
+/// Functionality related to diagnosing problems with a project or development
+/// environment.
+abstract class DoctorDomain extends Domain {
+
+  /// Diagnose problems with the user's environment.
```
Comment (zanderso) [inline packages/flutter_tool_api/lib/doctor.dart]: If this is the entrypoint of the doctor extension, it should document what is passed in `arguments` and what should be done with them.

File: packages/flutter_tool_api/lib/doctor.dart [hunk 6]
```diff
@@ -0,0 +1,223 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (zanderso) [inline packages/flutter_tool_api/lib/doctor.dart]: Here you could do:
```dart
/// This library contains the extension API for extending the 'flutter doctor' command with new checks.
/// Here is a simple example:
///
/// class MyDoctor extends DoctorDomain {
///   ...
/// }
///
/// For a fuller example see ...
library doctor;
```

File: packages/flutter_tool_api/pubspec.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
+name: flutter_tool_api
+description: API for building Flutter tool extensions
+homepage: https://flutter.dev
+author: Flutter Authors <flutter-dev@googlegroups.com>
+
+environment:
+  # The pub client defaults to an <2.0.0 sdk constraint which we need to explicitly overwrite.
+  sdk: ">=2.2.2 <3.0.0"
```
Comment (zanderso) [inline packages/flutter_tool_api/pubspec.yaml]: Why is the minimum requirement so high?

File: packages/flutter_tool_api/pubspec.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+
+dependencies:
+  file:
```
Comment (zanderso) [inline packages/flutter_tool_api/pubspec.yaml]: Is this like 'any' ?

File: packages/flutter_tool_api/lib/extension.dart [hunk 9]
```diff
@@ -0,0 +1,184 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
```
Comment (zanderso) [inline packages/flutter_tool_api/lib/extension.dart]: Ditto comment about a top-level library dartdoc comment.

File: packages/flutter_tool_macos/lib/main.dart [hunk 10]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (zanderso) [inline packages/flutter_tool_macos/lib/main.dart]: This might be a bit easier to bootstrap if this were kept in the flutter_tool package as a "builtin" extension. Could still spawn an isolate to run it, and disallow importing things it shouldn't import via tests.

File: packages/flutter_tools/lib/src/base/extension_host.dart [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+import '../convert.dart';
+import '../globals.dart';
+import 'file_system.dart';
+import 'platform.dart';
+
+/// Generates and caches a snapshot of the tool extension apis to apply.
+class CrossIsolateShim {
+  CrossIsolateShim() {
+    // HARDCODED.
+    const String path = '/Users/jonahwilliams/Documents/flutter/packages/flutter_tool_macos';
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/base/extension_host.dart]: It might help removing this hardcoding by bootstrapping with "builtin" extensions that live in the flutter_tool package.

File: packages/flutter_tools/lib/src/context_runner.dart [hunk 12]
```diff
@@ -102,7 +101,6 @@ Future<T> runInContext<T>(
       UserMessages: () => UserMessages(),
       VisualStudio: () => VisualStudio(),
       VisualStudioValidator: () => const VisualStudioValidator(),
-      WebWorkflow: () => const WebWorkflow(),
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/context_runner.dart]: This change seems unrelated

File: packages/flutter_tools/lib/src/doctor.dart [hunk 13]
```diff
@@ -334,6 +342,42 @@ abstract class DoctorValidator {
   Future<ValidationResult> validate();
 }
 
+/// A compatibility layer until all custom doctor implementations use the
+/// extension shim.
+class ExtensionCompatDoctorValidation implements DoctorValidator {
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/doctor.dart]: I think we'll need to change this:

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/doctor.dart#L252

From using runtimeType for the validator name to something else.

File: packages/flutter_tool_api/test/extension_test.dart [hunk 14]
```diff
@@ -0,0 +1,143 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
```
Comment (zanderso) [inline packages/flutter_tool_api/test/extension_test.dart]: Note that we'll need tests that the core tool fails gracefully when the extension isolates misbehave in various ways.

File: packages/flutter_tool_api/pubspec.yaml (hunk 7 above)
Comment (jonahwilliams) [reply packages/flutter_tool_api/pubspec.yaml]: No reason, just copy paste

File: packages/flutter_tool_api/pubspec.yaml (hunk 8 above)
Comment (jonahwilliams) [reply packages/flutter_tool_api/pubspec.yaml]: any for now, since `update-packages` will diff update everything, I am saving that for the end

**Comment by jonahwilliams** [comment]
Closing this for now, since I'm working on other things
